src/main/java/frc/robot/subsystems/swerve/PhoenixOdometryThread.java

```diff
@@ -51,88 +59,97 @@ public static PhoenixOdometryThread getInstance() {
     return instance;
   }
 
-  private PhoenixOdometryThread() {
+  // Used for testing
+  public static PhoenixOdometryThread createWithJournal(final Queue<Samples> journal) {
+    return new PhoenixOdometryThread(journal);
+  }
+
+  private PhoenixOdometryThread(final Queue<Samples> journal) {
     setName("PhoenixOdometryThread");
     setDaemon(true);
+
+    this.journal = journal;
   }
 
-  @Override
-  public void start() {
-    if (timestampQueues.size() > 0) {
-      super.start();
-    }
+  private PhoenixOdometryThread() {
+    this(EvictingQueue.create(20));
   }
 
-  public Queue<Double> registerSignal(ParentDevice device, StatusSignal<Double> signal) {
-    Queue<Double> queue = new ArrayDeque<>(100);
-    signalsLock.lock();
-    SwerveSubsystem.odometryLock.lock();
+  public void registerSignals(Collection<Registration> registrations) {
+    registerSignals(registrations.toArray(new Registration[] {}));
+  }
+
+  // Returns a handle which can be used to collect the last 20 signal results
+  public void registerSignals(Registration... registrations) {
+    var writeLock = journalLock.writeLock();
+
     try {
-      isCANFD = CANBus.isNetworkFD(device.getNetwork());
-      BaseStatusSignal[] newSignals = new BaseStatusSignal[signals.length + 1];
-      System.arraycopy(signals, 0, newSignals, 0, signals.length);
-      newSignals[signals.length] = signal;
-      signals = newSignals;
-      queues.add(queue);
+      writeLock.lock();
+
+      for (var registration : registrations) {
+        assert CANBus.isNetworkFD(registration.device.getNetwork()) : "Only CAN FDs supported";
+
+        signals.addAll(registration.signals);
+      }
     } finally {
-      signalsLock.unlock();
-      SwerveSubsystem.odometryLock.unlock();
+      writeLock.unlock();
     }
-    return queue;
   }
 
-  public Queue<Double> makeTimestampQueue() {
-    Queue<Double> queue = new ArrayDeque<>(100);
-    SwerveSubsystem.odometryLock.lock();
+  public List<Samples> samplesSince(double timestamp, Set<StatusSignal<Double>> signals) {
+    var readLock = journalLock.readLock();
     try {
-      timestampQueues.add(queue);
+      readLock.lock();
+
+      return journal.stream()
+          .filter(s -> s.timestamp > timestamp)
+          .map(
+              s -> {
+                var filteredValues =
+                    s.values.entrySet().stream()
+                        .filter(e -> signals.contains(e.getKey()))
+                        .collect(
+                            Collectors.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+                return new Samples(s.timestamp, filteredValues);
```

Comment:
any gc pressure concerns here?

Reply:
Not really. These are short lived objects that shouldn't make it out of the gc's eden. We're also limited to 20 samples stored in the evicting queue at the moment, so we're filtering over 20*signals.size objects total. Should be pretty slight in our use. Always better to measure than infer though, if we see issues.